lib/bloc/account/account_bloc.dart

```diff
@@ -663,28 +663,36 @@ class AccountBloc {
     DateTime _firstDate;
     print("refreshing payments...");
 
-    return _breezLib.getPayments().then((payments) {
-      List<PaymentInfo> _paymentsList = payments.paymentsList.map((payment) {
-        var singlePaymentInfo =
-            SinglePaymentInfo(payment, _accountController.value);
-
-        return singlePaymentInfo;
-      }).toList();
+    return _breezLib.getPayments().then((payments) async {
+      List<PaymentInfo> _paymentsList =
+          await Stream.fromIterable(payments.paymentsList)
+              .asyncMap((payment) => _fetchPaymentInfo(payment))
+              .toList();
 
       if (_paymentsList.length > 0) {
         _firstDate = DateTime.fromMillisecondsSinceEpoch(
-            _paymentsList.last.creationTimestamp.toInt() * 1000);
+          _paymentsList.last.creationTimestamp.toInt() * 1000,
+        );
       }
       print("refresh payments finished " +
           payments.paymentsList.length.toString());
       return PaymentsModel(
-          _paymentsList,
-          _filterPayments(_paymentsList),
-          _paymentFilterController.value,
-          _firstDate ?? DateTime(DateTime.now().year));
+        _paymentsList,
+        _filterPayments(_paymentsList),
+        _paymentFilterController.value,
+        _firstDate ?? DateTime(DateTime.now().year),
+      );
     });
   }
 
+  Future<PaymentInfo> _fetchPaymentInfo(Payment payment) async {
+    return SinglePaymentInfo(
+      payment,
+      _accountController.value,
+      await _posRepository.fetchSaleByPaymentHash(payment.paymentHash),
```

Comment:
This will hit the sql db on every payment. Instead we can query once for all sales items, create a hash table in memory and use it to fetch a sale by hash.

Reply:
Changes applied ;)